dashboard/app/controllers/script_levels_controller.rb

```diff
@@ -247,15 +247,48 @@ def stage_extras
       return
     end
 
-    @stage = Script.get_from_cache(params[:script_id]).lesson_by_relative_position(params[:stage_position].to_i)
+    user = @user || current_user
+    @stage = Script.get_from_cache(
+      params[:script_id]
+    ).lesson_by_relative_position(
+      params[:stage_position].to_i
+      )
     @script = @stage.script
+    script_bonus_levels_by_stage = @script.get_bonus_script_levels(@stage)
+
+    # bonus level summaries are cached, so explicitly don't contain any
+    # user-specific data. hence we need to merge the user's progress and
+    # localized level display name and description into the cached data.
+    script_bonus_levels_by_stage.each do |stage|
+      stage[:levels].each do |level_summary|
+        ul = UserLevel.find_by(
+          level_id: level_summary[:level_id], user_id: user.id, script: @script
+        )
+        level_summary[:perfect] = ul&.perfect?
+        localized_level_description = I18n.t(
+          ul.level.name,
+          scope: [:data, :bubble_choice_description],
+          default: ul.level.bubble_choice_description
+        )
+        localized_level_display_name = I18n.t(
```

Comment:
Sorry to lurk here but I'm wondering why you didn't add the locale to the cache key and keep the localization logic where it was? It seems to make more sense in the model (to me) if possible

Reply:
thank you for lurking! i have no idea what you're talking about!

Reply:
that said, i would like to push back on this making more sense in the model. `ScriptLevel` and `UserLevel` are separate models, so imo a summary of one shouldn't include anything from the other.

Reply:
sorry i was mixing up the user data issue with the localization issue. i don't think we should be including user data in the script level summary, but including localization in the model makes sense _if we have locale-based caching_... which i don't believe we have yet.

Reply:
not sure if this got cleared up yesterday, but we do have locale-based caching, configured here: https://github.com/code-dot-org/code-dot-org/blob/staging/cookbooks/cdo-varnish/libraries/http_cache.rb if the language cookie is allowed through, then the page can be cached differently for each locale.

Reply:
@davidsbailey Are the language cookie (and header) the _only_ ways to set locale in our system?  (If not, we would still need to avoid caching language-specific objects in dashboard.)

Reply:
> Are the language cookie (and header) the only ways to set locale in our system?

my understanding is yes, but the other reviewers may have more accurate info.

Reply:
I'm not seeing why a lesson extras path in the express course would be cached at all, though, because I do not see express listed in the `CACHED_SCRIPTS_MAP`.

Reply:
http_cache.rb controls which cookies and headers are allowed through which affect the granularity with which we cache things. the other half of the puzzle is what cache headers get set / whether the page is allowed to be cached at all. if express is not supposed to be cached, then it seems like we actually shouldn't be allowing caching. so it could be worth checking the cache headers, and/or seeing if this logic is doing the right thing for lesson extras: https://github.com/code-dot-org/code-dot-org/blob/97de9db301efdf8d4bdcf967bb560c7b7a4bf0f9/dashboard/app/controllers/script_levels_controller.rb#L24-L28

Reply:
Just to clarify, the cache that Charlie is talking about here is the script cache in dashboard (not http caching).

Reply:
ahh, thanks James! Sorry, I misunderstood that. most of what I said here is off topic, then.